src/Middleware/HttpLogging/src/HttpLoggingMiddleware.cs

```diff
@@ -0,0 +1,289 @@
+                    list.Add(new KeyValuePair<string, object?>(nameof(response.Body), body));
+                }
+
+                var httpResponseLog = new HttpResponseLog(list);
+
+                _logger.Log(LogLevel.Information,
+                    eventId: LoggerEventIds.ResponseLog,
+                    state: httpResponseLog,
+                    exception: null,
+                    formatter: HttpResponseLog.Callback);
+            }
+            finally
+            {
+                if (_options.LoggingFields.HasFlag(HttpLoggingFields.ResponseBody))
+                {
+                    if (bufferingStream != null)
+                    {
+                        bufferingStream.Dispose();
+                    }
+
+                    context.Features.Set(originalBodyFeature);
+                }
+            }
+        }
+
+        private bool IsSupportedMediaType(string contentType)
+        {
+            var mediaTypeList = _options.SupportedMediaTypes;
+            if (mediaTypeList == null || mediaTypeList.Count == 0 || string.IsNullOrEmpty(contentType))
+            {
+                return false;
+            }
+
+            var mediaType = new MediaTypeHeaderValue(contentType);
+            foreach (var type in mediaTypeList)
+            {
+                if (mediaType.IsSubsetOf(type))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private void FilterHeaders(List<KeyValuePair<string, object?>> keyValues, IHeaderDictionary headers, ISet<string> allowedHeaders)
+        {
+            foreach (var (key, value) in headers)
+            {
+                if (!allowedHeaders.Contains(key))
+                {
+                    // Key is not among the "only listed" headers.
+                    keyValues.Add(new KeyValuePair<string, object?>(key, "X"));
+                    continue;
+                }
+                keyValues.Add(new KeyValuePair<string, object?>(key, value.ToString()));
+            }
+        }
+
+        private async Task<string> ReadRequestBody(HttpRequest request, CancellationToken token)
+        {
+            if (_options.BodyEncoding == null)
+            {
+                return "X";
+            }
+
+            using var joinedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(token);
+            joinedTokenSource.CancelAfter(_options.RequestBodyTimeout);
+            var limit = _options.RequestBodyLogLimit;
+
+            // Use a pipe for smaller payload sizes
+            if (limit <= PipeThreshold)
+            {
+                try
+                {
+                    while (true)
+                    {
+                        // TODO if someone uses the body after this, it will not have the rest of the data.
+                        var result = await request.BodyReader.ReadAsync(joinedTokenSource.Token);
```

Comment:
@davidfowl @halter73 @Tratcher I tried to get fancy here with request body reading, but this won't work if the middleware above it uses the Stream Body instead of the BodyReader. So I may need to always call enable buffering.

Reply:
Here's a question for you. Why isn't this a decorator PipeReader/Stream instead of one that reads up front? Is the assumption that we want to read bodies up front this with middleware?

Reply:
>  Is the assumption that we want to read bodies up front this with middleware?

At the moment yes

> Why isn't this a decorator PipeReader/Stream instead of one that reads up front?
Because by the time we are in the middleware, none of the request body may have actually been read. So either we need to read here, or we need to add the log to the stream itself and log there when the body is read.

There are tradeoffs to both; it may be better to add a wrapper, but that means the request body log will no happen in the middleware at all and at some arbitrary point.

Reply:
We have to go the wrapper route, reading the body up front breaks scenarios like auth+100-continue, per endpoint size and rate limits, etc. We also want to avoid buffering the whole body, it should be logged in chunks.

Reply:
> We have to go the wrapper route, reading the body up front breaks scenarios like auth+100-continue, per endpoint size and rate limits, etc.

I don't think I quite understand how this breaks. Care to explain more? I don't see how this is different than any middleware or user who calls EnableBuffering.

> We also want to avoid buffering the whole body, it should be logged in chunks.

I'm still against this, I think the user experience of having a single log message

Reply:
EnableBuffering wraps, it doesn't read until the app reads. The app still has time to change parameters on the request like the size or rate limits, or reject the request without reading the body such as a 401. To be clear this is a ship blocker, the logging middleware must not break other functionality in the app.

We want to limit the side-effects this middleware has based on where it's placed in the pipeline. A middleware like this wants to be placed early in the pipeline to capture the most requests, but it still wants to honor things like per-request size limits and auth from endpoints. The best way to do that is to let the app/pipeline control when the body gets read.

32k is ok to buffer, but what happens when someone turns that limit up to 32mb, or 1gb? We want to be able to stream data to the logs in reasonable chunks without holding onto all of it. Having everything in one log is nice, but it does not scale.

Reply:
>EnableBuffering wraps, it doesn't read until the app reads. The app still has time to change parameters on the request like the size or rate limits, or reject the request without reading the body such as a 401

Ah I see, that does break things. Looks like we gotta change it to log when we have enough of the request and response body then.